dg_test.go

```diff
@@ -715,46 +720,30 @@ func TestDirectedGraph_TestWaitingResolution(t *testing.T) {
 	err = dependencyNode1.ResolveNode(dgraph.Waiting)
 	assert.NoError(t, err)
 	// It's waiting, still, so nothing should resolve.
-	readyNodes = d.PopReadyNodes()
-	assert.Equals(t, len(readyNodes), 0)
+	assert.Equals(t, d.HasReadyNodes(), false)
 }
 
-// This test is here just to validate that multi-threaded usage does not cause any problems.
-// Recommended to run with `-race` and `-count 10000`
-func TestDirectedGraph_TwoOrDependenciesMultiThreaded(t *testing.T) {
+func TestDirectedGraph_PushStartingNodes(t *testing.T) {
 	d := dgraph.New[string]()
-	dependentNode, err := d.AddNode("dependent-node", "Dependent Node")
+	noDependencies, err := d.AddNode("no-dependencies", "No Dependencies")
 	assert.NoError(t, err)
-	dependencyNode1, err := d.AddNode("dependency-node-1", "Dependency 1")
+	onlyObviatedDependencies, err := d.AddNode("only-obviated-dependencies", "Only Obviated Dependencies")
 	assert.NoError(t, err)
-	dependencyNode2, err := d.AddNode("dependency-node-2", "Dependency 2")
+	withANDDependencies, err := d.AddNode("with-and-dependencies", "With AND Dependencies")
+	assert.NoError(t, err)
+	withORDependencies, err := d.AddNode("with-or-dependencies", "With OR Dependencies")
 	assert.NoError(t, err)
 
-	wg := sync.WaitGroup{}
-	wg.Add(2)
-	go func() {
-		defer wg.Done()
-		assert.NoError(t, dependentNode.ConnectDependency(dependencyNode1.ID(), dgraph.OrDependency))
-	}()
-	go func() {
-		defer wg.Done()
-		assert.NoError(t, dependentNode.ConnectDependency(dependencyNode2.ID(), dgraph.OrDependency))
-	}()
-	wg.Wait()
-
+	assert.NoError(t, withANDDependencies.ConnectDependency(noDependencies.ID(), dgraph.AndDependency))
+	assert.NoError(t, withANDDependencies.ConnectDependency(onlyObviatedDependencies.ID(), dgraph.AndDependency))
+	assert.NoError(t, withORDependencies.ConnectDependency(noDependencies.ID(), dgraph.OrDependency))
+	assert.NoError(t, withORDependencies.ConnectDependency(onlyObviatedDependencies.ID(), dgraph.OrDependency))
+	assert.NoError(t, onlyObviatedDependencies.ConnectDependency(noDependencies.ID(), dgraph.ObviatedDependency))
 	assert.NoError(t, d.PushStartingNodes())
-	wg = sync.WaitGroup{}
-	wg.Add(2)
-	go func() {
-		defer wg.Done()
-		assert.NoError(t, dependencyNode1.ResolveNode(dgraph.Resolved))
-	}()
-	go func() {
-		defer wg.Done()
-		readyStatus := dependencyNode2.IsReady()
-		resolutionStatus := dependencyNode1.ResolutionStatus()
-		t.Logf("Raced async ready status: %t", readyStatus)
-		t.Logf("Raced async resolution status: %s", resolutionStatus)
-	}()
-	wg.Wait()
+	assert.Equals(t, d.HasReadyNodes(), true)
+	readyNodes := d.PopReadyNodes()
+	assert.Equals(t, len(readyNodes), 2)
+	assert.MapContainsKey(t, noDependencies.ID(), readyNodes)
+	// Obviated dependencies should not count, so it should have been marked ready.
+	assert.MapContainsKey(t, onlyObviatedDependencies.ID(), readyNodes)
```

Comment:
The code at line 748 is nicely expressive -- it pretty clearly says that a node which only has obviated dependencies should be ready -- so I think we can dispense with the comment at line 747.

Reply:
Sure.